operators/advanced_network/python/check_system.py

```diff
@@ -0,0 +1,632 @@
+            "Could not determine online CPUs. File /sys/devices/system/cpu/online not found."
+        )
+        sys.exit(1)
+
+
+def check_cpu_governor():
+    """
+    Checks if the CPU frequency governor is set to 'performance' for all online CPUs.
+    """
+    online_cpus = get_online_cpus()
+
+    for cpu in online_cpus:
+        scaling_governor_path = f"/sys/devices/system/cpu/cpu{cpu}/cpufreq/scaling_governor"
+
+        try:
+            with open(scaling_governor_path, "r") as f:
+                governor = f.read().strip()
+
+            if governor == "performance":
+                logging.info(f"CPU {cpu}: Governor is correctly set to 'performance'.")
+            else:
+                logging.warning(f"CPU {cpu}: Governor is set to '{governor}', not 'performance'.")
+
+        except FileNotFoundError:
+            logging.error(
+                f"CPU {cpu}: Scaling governor file not found. This CPU may not support frequency scaling."
+            )
+        except PermissionError:
+            logging.error(
+                f"CPU {cpu}: Permission denied while accessing scaling governor file. Run as root."
+            )
+
+
+def check_mrrs():
+    """
+    Checks if the Maximum Read Request Size (MRRS) of NVIDIA Ethernet controllers
+    is set to 4096.
+    """
+    try:
+        _, addrs = get_nic_info()
+
+        for pci_address in addrs:
+
+            # Query MRRS for the NIC using setpci
+            mrrs_result = subprocess.run(
+                ["setpci", "-s", pci_address, "68.w"], capture_output=True, text=True, check=True
+            )
+
+            # Convert MRRS value from hexadecimal to decimal
+            mrrs_value = (int(mrrs_result.stdout.strip(), 16) & 0xF000) >> 12
+
+            if mrrs_value == 5:
+                logging.info(f"{pci_address}: MRRS is correctly set to 4096.")
+            else:
+                logging.warning(f"{pci_address}: MRRS is set to {2**(7+mrrs_value)}, not 4096.")
+
+    except FileNotFoundError:
+        logging.error("The required tools (lspci or setpci) are not available on this system.")
+        sys.exit(1)
+    except subprocess.CalledProcessError as e:
+        logging.error(f"Error while checking MRRS: {e}")
+        sys.exit(1)
+
+
+def check_max_payload_size():
+    """
+    Checks the Maximum Payload Size (MPS) of NVIDIA Ethernet controllers
+    from the DevCtl section and ensures it is set to 256 bytes.
+    """
+    try:
+        _, addrs = get_nic_info()
+
+        for pci_address in addrs:
+            # Query detailed device information using lspci -vv
+            mps_result = subprocess.run(
+                ["lspci", "-vv", "-s", pci_address], capture_output=True, text=True, check=True
+            )
+
+            # Parse MaxPayload information from the DevCtl section
+            lines = mps_result.stdout.splitlines()
+            devctl_found = False
+            for i, line in enumerate(lines):
+                if "DevCtl:" in line:
```

Comment:
I don't have this for mine:

```
lspci -vv -s 0005:03:00
0005:03:00.0 Ethernet controller: Mellanox Technologies MT2910 Family [ConnectX-7]
        Subsystem: Mellanox Technologies MT2910 Family [ConnectX-7]
        Control: I/O- Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx+
        Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
        Latency: 0
        Interrupt: pin A routed to IRQ 195
        IOMMU group: 6
        Region 0: Memory at 2b00000000 (64-bit, prefetchable) [size=32M]
        Expansion ROM at 2e31800000 [virtual] [disabled] [size=1M]
        Capabilities: <access denied>
        Kernel driver in use: mlx5_core
        Kernel modules: mlx5_core

0005:03:00.1 Ethernet controller: Mellanox Technologies MT2910 Family [ConnectX-7]
        Subsystem: Mellanox Technologies MT2910 Family [ConnectX-7]
        Control: I/O- Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx+
        Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
        Latency: 0
        Interrupt: pin B routed to IRQ 195
        IOMMU group: 6
        Region 0: Memory at 2b02000000 (64-bit, prefetchable) [size=32M]
        Expansion ROM at 2e31900000 [virtual] [disabled] [size=1M]
        Capabilities: <access denied>
        Kernel driver in use: mlx5_core
        Kernel modules: mlx5_core
```

And obviously I've had no issues running the ANO app. Not sure how we should deal with such errors when we see that it's not actually necessary.

Reply:
You probably need sudo

Reply:
I will add a check to make sure they run it with root. There's no way you can get -vv information from lspci that we need without root afaik.